library/asn1parse.c

```diff
@@ -98,6 +98,11 @@ int mbedtls_asn1_get_len(unsigned char **p,
     return 0;
 }
 
+static int mbedtls_asn1_bad_tag(int tag)
+{
+    return tag < 1 || (tag & ~0xE0) >= 0x1F;
```

Comment:
`tag` is always ≤255, so `(tag & ~0xE0) >= 0x1F` can't happen. `tag` is always ≥0 so `tag < 1` is equivalent to `tag == 0`.
```suggestion
    return tag == 0;
```

Reply:
`(tag & ~0xE0) == 0x1F` can indeed happen (consider `tag == 0xFF`).  Also `tag < 1` should be `(tag & 0xDF) == 0`.

Reply:
Ah, my bad, I'd reasoned as if it was `> 0x1F` rather than `>= 0x1F`. It would indeed have been more readable with `==`. And please be consistent with `0xDF` vs `0xE0`. In general, please avoid negation when not needed, especially if it leads to double negation. A `is_good_tag` function is likely to make the code more readable than `is_bad_tag`.